### PR TITLE
Add exception handling around fapd handle is_active() call

### PR DESCRIPTION
**Pull request by tparchambault**

### CODE REVIEW

**Comment by tparchambault** [comment]
After merging master into this branch, on a clean FC34 VM (installed emacs, fapolicyd, the generated fapolicy-analyzer rpm), w/fapd running in `--permissive` after adding a number of files to the ancillary trust dbase and attempting a deploy, the following was observed:

```
 DEBUG:root:SessionManager::__autosave_edit_session()
DEBUG:root: Session autosave is disabled/bypassed
DEBUG:root:Deploying...
DEBUG:root:fapd_dbase::fapd_dbase_snapshot()
 Creating '/root/.local/share/fapolicy-analyzer/' 
DEBUG:root:Xdg absolute path: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup
Fapolicyd backup to: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_145633_497756.tgz
DEBUG:root:fapd_dbase_cleanup_snapshots()
DEBUG:root:Search Pattern: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_*.tgz
DEBUG:root:Glob search results: ['/root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_145633_497756.tgz']
DEBUG:root:_status(ServiceStatus.FALSE updated
DEBUG:root:monitor_daemon:Dispatch update request
DEBUG:root:on_update_daemon_status(ServiceStatus.FALSE)
ERROR:root:An error occurred trying to deploy the changes. Please try again.: DbusFailure(D-Bus error: Transaction for fapolicyd.service/start is destructive (fapolicyd.service has 'stop' job queued, but 'start' is included in transaction). (org.freedesktop.systemd1.TransactionIsDestructive))
DEBUG:root:disposing of subscription for class DeployChangesetsOp
DEBUG:root:_update_fapd_status(ServiceStatus.FALSE)
```

**Comment by tparchambault** [comment]
With the monitoring thread not running,

```
DEBUG:root:Starting the fapd monitoring thread
DEBUG:root:fapd monitoring thread commented out

```

we still observe the full assortment of errors wrt restarting the daemon after stopping it.
Here, at deployment:

```
DEBUG:root:Deploying...
DEBUG:root:fapd_dbase::fapd_dbase_snapshot()
DEBUG:root:Xdg absolute path: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup
Fapolicyd backup to: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_184752_502467.tgz
DEBUG:root:fapd_dbase_cleanup_snapshots()
DEBUG:root:Search Pattern: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_*.tgz
DEBUG:root:Glob search results: ['/root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_175559_584055.tgz', '/root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_182913_127940.tgz', '/root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_183347_403512.tgz', '/root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_184752_502467.tgz']
DEBUG:root:Deleting: /root/.local/share/fapolicy-analyzer/Fapolicyd_Backup_20220727_175559_584055.tgz
ERROR:root:An error occurred trying to deploy the changes. Please try again.: DbusFailure(D-Bus error: Transaction for fapolicyd.service/start is destructive (fapolicyd.service has 'stop' job queued, but 'start' is included in transaction). (org.freedesktop.systemd1.TransactionIsDestructive))
DEBUG:root:disposing of subscription for class DeployChangesetsOp
```

Also occurs on rollback, but the DBus error message does not always get displayed in the logging.debug() messages. They will show up in the journal though.

```
Jul 27 19:11:45 fedora systemd[1]: Stopping File Access Policy Daemon...
Jul 27 19:11:46 fedora systemd[1]: Requested transaction contradicts existing jobs: Transaction for fapolicyd.service/start is destructive (fapolicyd.service has 'stop' job queued, but 'start' is included in transaction).
Jul 27 19:11:48 fedora systemd[1]: fapolicyd.service: Deactivated successfully.
Jul 27 19:11:48 fedora systemd[1]: Stopped File Access Policy Daemon.
```
The issue is always the daemon start operation however there is no service start entry in the journal so possibly the fapolicyd start request is not being delivered to systemd or it's delivered in a format that is inconsistent/nonsensical from systemd's perspective based on the `transaction contradicts...` log msg.

**Comment by tparchambault** [comment]
From the perspective of systemd, bringing down fapolicyd takes `~3 secs` and in that 3 sec window, systemd is getting a request to restart fapolicyd:

```
Jul 27 20:30:03 fedora systemd[1]: Stopping File Access Policy Daemon...
Jul 27 20:30:03 fedora fapolicyd[1698]: shutting down...
Jul 27 20:30:04 fedora systemd[1]: Requested transaction contradicts existing jobs: Transaction for fapolicyd.service/start is destructive (fapolicyd.service has 'stop' job queued, but 'start' is included in transaction).
Jul 27 20:30:06 fedora systemd[1]: fapolicyd.service: Deactivated successfully.
Jul 27 20:30:06 fedora audit[1]: SERVICE_STOP pid=1 uid=0 auid=4294967295 ses=4294967295 subj=system_u:system_r:init_t:s0 msg='unit=fapolicyd comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
Jul 27 20:30:06 fedora systemd[1]: Stopped File Access Policy Daemon.
Jul 27 20:30:06 fedora systemd[1]: fapolicyd.service: Consumed 1.306s CPU time.
```

**Comment by tparchambault** [comment]
Installed and tested over FC34 and Rhel 8.6, both ancillary trust changeset deployments and rollbacks. Tested with fapd in online and permissive modes. Good to go.

File: fapolicy_analyzer/ui/fapd_manager.py [hunk 1]
```diff
@@ -209,15 +210,16 @@ def _status(self):
             logging.debug("fapd is currently DISABLED")
             return ServiceStatus.UNKNOWN
         elif self.mode == FapdMode.ONLINE:
-            try:
-                if self._fapd_lock.acquire(blocking=False):
+            if self._fapd_lock.acquire(blocking=False):
```
Comment (jw3) [inline fapolicy_analyzer/ui/fapd_manager.py:213]: Can you use a `with` block here?

File: fapolicy_analyzer/ui/fapd_manager.py [hunk 2]
```diff
@@ -227,10 +229,15 @@ def _status(self):
                 return ServiceStatus.FALSE
 
     def capture_online_state(self):
-        logging.debug("capture_online_state()")
-        updated_online_state = ServiceStatus(self._fapd_ref.is_active())
-        self._fapd_prior_online_state = updated_online_state
-        self._fapd_cur_online_state = updated_online_state
+        logging.debug("capture_online_state(): Capturing current online status")
+        if self._fapd_lock.acquire():
```
Comment (jw3) [inline fapolicy_analyzer/ui/fapd_manager.py:233]: Same here, `with` would be ideal on these locked blocks.

File: fapolicy_analyzer/ui/fapd_manager.py [hunk 3]
```diff
@@ -241,9 +248,14 @@ def revert_online_state(self):
 
     def initial_daemon_status(self):
         if self._fapd_lock.acquire():
-            self._fapd_ref = Handle("fapolicyd")
-            if self._fapd_ref.is_valid():
-                self._fapd_status = ServiceStatus(self._fapd_ref.is_active())
-            else:
-                self._fapd_status = ServiceStatus.UNKNOWN
+            try:
```
Comment (jw3) [inline fapolicy_analyzer/ui/fapd_manager.py:251]: ... and again.

**Review by jw3 — CHANGES_REQUESTED** [review]
Using context blocks on the locks would be nice.
- https://docs.python.org/3.6/library/threading.html#with-locks

I can also make this fail by 
1. Start analyzer
2. Stop daemon
3. Remove the fapolicyd.service file
4. Start daemon from System menu
5. Observe the GUI is locked

The stack trace from the above steps
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/fapolicy_analyzer/ui/main_window.py", line 417, in on_fapdStartMenu_activate
    self._fapd_mgr.start()
  File "/usr/lib/python3.6/site-packages/fapolicy_analyzer/ui/fapd_manager.py", line 103, in start
    pid = self._start(instance)
  File "/usr/lib/python3.6/site-packages/fapolicy_analyzer/ui/fapd_manager.py", line 138, in _start
    self._fapd_ref.start()
RuntimeError: DbusFailure(D-Bus error: Failed to get file context on /usr/lib/systemd/system/fapolicyd.service. (org.freedesktop.DBus.Error.AccessDenied))
```

**Comment by jw3** [comment]
@tparchambault There are a few other spots that lock release needs to be ensured, looks like five spots total.  I think that using `with` at all of these is appropriate, I just can't mark them in the review because they are aren't part of your changes.

**Comment by tparchambault** [comment]
I'll try and research this more... I didn't go with the Context Manager approach because I did not see clear documentation when try/exception blocks were involved. We may be able to get rid of one or the other.

File: fapolicy_analyzer/ui/fapd_manager.py (hunk 1 above)
Comment (tparchambault) [reply fapolicy_analyzer/ui/fapd_manager.py:213]: I looked into that approach originally. The fly/ointment thing is the blocking=False argument, i.e. if the lock is not available then don't wait and return True. Fapd control operations (start/stop) are given priority over the never-ending periodic status requests.

Still digging around. I would still prefer to have locking somewhere given the status requests and stop/start requests originate on differing threads. I don't care if it's under the hood in rust-land or not.

File: fapolicy_analyzer/ui/fapd_manager.py (hunk 2 above)
Comment (tparchambault) [reply fapolicy_analyzer/ui/fapd_manager.py:233]: Yeah, `with` context management is Ok here. Will do.

**Comment by tparchambault** [comment]
Added context manager blocks to manage Lock acquire()/release() ops. Non-blocking Lock.acquire() left as originally coded.

**Comment by jw3** [comment]
Looks improved.  I can't make it fail by moving the unit file around while the app is up.

Only issue I see is that failures do not make it to the UI.  Is it possible to use a toaster message on these failures?

**Review by jw3 — APPROVED** [review]
The toasters would be nice, but not blocking.

**Comment by tparchambault** [comment]
> Only issue I see is that caught exceptions do not send any notification to the UI. Is it possible to use a toaster message on these failures?

Sure I'll look at that tomorrow AM.

**Comment by jw3** [comment]
Let's add the toaster for the start/stop controls in this PR.

For the status checks let's have a separate issue that instead of a toaster we get a third icon on the status display.  For this PR however lets just log the failure to stderr.

**Pull request merged**